ai/core/src/unitycatalog/ai/core/utils/callable_utils.py

```diff
@@ -769,3 +772,266 @@ def check_docstring_signature_consistency(
             UserWarning,
             stacklevel=2,
         )
+
+
+def _recursively_parse_type(type_info: Any) -> str:
+    """
+    Recursively parse a SQL type definition (provided either as a dict or a string)
+    into a Python type annotation string.
+
+    Args:
+        type_info: A dict containing the type information, or a simple string.
+
+    Returns:
+        A string representing the Python type.
+    """
+    if not isinstance(type_info, dict):
+        mapped = SQL_TYPE_TO_PYTHON_TYPE_MAPPING.get(str(type_info).upper(), None)
+        # NB: OSS Unity Catalog doesn't retain the inner subtypes for collections
+        # so the logic below for handling arrays and maps is unnecessary.
+        # We can only retrieve the outer type and will short circuit early here.
+        if isinstance(mapped, tuple):
+            mapped = mapped[0]
+        return mapped.__name__ if mapped else "Any"
+
+    t = type_info.get("type")
+    if not t:
+        return "Any"
+    t_lower = str(t).lower()
+
+    if t_lower == "array":
+        element = type_info.get("elementType")
+        return f"list[{_get_mapped_value(element)}]"
+    elif t_lower == "map":
+        key = type_info.get("keyType")
+        value = type_info.get("valueType")
+        return f"dict[{_get_mapped_value(key)}, {_get_mapped_value(value)}]"
+    elif t_lower == "struct":
+        return "dict"
+    else:
+        mapped = SQL_TYPE_TO_PYTHON_TYPE_MAPPING.get(str(t).upper(), None)
+        return mapped.__name__ if mapped else "Any"
+
+
+def _get_mapped_value(x: Any) -> str:
+    """
+    Helper function to get the Python type as a string for a given SQL type.
+    If x is a dict, it recursively parses it; otherwise, it maps the type via SQL_TYPE_TO_PYTHON_TYPE_MAPPING.
+    """
+    if isinstance(x, dict):
+        return _recursively_parse_type(x)
+    else:
+        mapped = SQL_TYPE_TO_PYTHON_TYPE_MAPPING.get(str(x).upper(), None)
```

Comment:
Should we have the tuple check here as well? 
Also seems like we don't need this function, wherever calls `_get_mapped_value` can just call `_recursively_parse_type`